### ISSUE TITLE
StreamQueue hasNext property is not intuitive.

### ISSUE THREAD

**Issue by cvanvlack** [dart-lang/async#205]
The following code does not behave as one would expect and the documentation does not make it clear what one should expect from the `hasNext` property
```
late StreamController<int> streamController = StreamController<int>();
late StreamQueue<int> streamQueue = StreamQueue<int>(streamController.stream);
bool hasNext = await streamQueue.hasNext;
```
This code will wait indefinitely rather than returning false. Having a `hasNext` property that never returns when there is no `next` is very difficult to use.

**Comment by lrhn** [comment]
That's just how streams work.
You cannot know whether there is a `next` element unless the stream is closed (then there wasn't).
Until the stream closes, there might come a next event at any later time. 
*When* would `hasNext` return false?

The only way for `hasNext` to say "no", other than waiting for the stream to close, is to have a timeout.
That's a possible addition, but not something inherently necessary to the current operation.

In other words, this is working as intended, and doing anything else would be a very different operation than the current `hasNext`.

**Comment by cvanvlack** [comment]
When I am using a StreamQueue, the implication is that I am using an object that is derived from a stream but behaves like a Queue.

I completely acknowledge my lack of experience with streams. But this is why I think the documentation is unclear. 

As a new Flutter developer learning these tools the documentation assumes knowledge of the implementation of other classes (Streams in this case). I don't know how this has been created. I just want clarity on what to expect from this class.

If the documentation stated "The future will never complete if the stream is not closed and there are no events." that would have saved me a bunch of time trying to figure out what I was doing wrong.

**Comment by lrhn** [comment]
That *is* a reasonable addition to the comments. 

Futures which never complete are always tricky to work with. You'd get the same effect by just doing `await stream.first` on a stream with no events and which doesn't close, so this is not in any way unique to `StreamQueue`.

A stream queue is not really a `Queue` - not in the sense that you both put data into one end and take them out the other.
Instead the stream is putting data into one end of the structure and you take them out from the other end. I guess it's more like a buffer. Naming is hard.
